pkg/controllers/endpointslicedns/cache.go

```diff
@@ -75,13 +90,18 @@ func (d *DNSCache) DeleteByResourceKey(resourceKey string) {
 		return
 	}
 	if fqdnToDelete, ok := d.resourceKeyToFQDN[resourceKey]; ok {
-		delete(d.entries, fqdnToDelete)
+		for i, entry := range d.entries[fqdnToDelete] {
+			if entry.ResourceKey == resourceKey {
+				d.entries[fqdnToDelete] = append(d.entries[fqdnToDelete][:i], d.entries[fqdnToDelete][i+1:]...)
+				break
+			}
```

Comment:
Should we add logic here to also delete the FQDN if all resources associated are removed?